br/pkg/lightning/restore/restore.go

```diff
@@ -1236,7 +1256,91 @@ func (rc *Controller) buildRunPeriodicActionAndCancelFunc(ctx context.Context, s
 
 var checksumManagerKey struct{}
 
+const (
+	pauseGCTTLForDupeRes      = time.Hour
+	pauseGCIntervalForDupeRes = time.Minute
+)
+
+func (rc *Controller) keepPauseGCForDupeRes(ctx context.Context) error {
+	tlsOpt := rc.tls.ToPDSecurityOption()
+	pdCli, err := pd.NewClientWithContext(ctx, []string{rc.cfg.TiDB.PdAddr}, tlsOpt)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer pdCli.Close()
+
+	serviceID := "lightning-duplicate-resolution-" + uuid.New().String()
+	ttl := int64(pauseGCTTLForDupeRes / time.Second)
+
+	var (
+		safePoint uint64
+		paused    bool
+	)
+	// Try to get the minimum safe point across all services as our GC safe point.
+	for i := 0; i < 10; i++ {
+		if i > 0 {
+			time.Sleep(time.Second * 3)
+		}
+		minSafePoint, err := pdCli.UpdateServiceGCSafePoint(ctx, serviceID, ttl, 1)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		newMinSafePoint, err := pdCli.UpdateServiceGCSafePoint(ctx, serviceID, ttl, minSafePoint)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if newMinSafePoint <= minSafePoint {
+			safePoint = minSafePoint
+			paused = true
+			break
+		}
+		log.L().Warn(
+			"Failed to register GC safe point because the current minimum safe point is newer"+
+				" than what we assume, will retry newMinSafePoint next time",
+			zap.Uint64("minSafePoint", minSafePoint),
+			zap.Uint64("newMinSafePoint", newMinSafePoint),
+		)
+	}
+	if !paused {
+		return errors.New("failed to pause GC for duplicate resolution after all retries")
+	}
+
+	go func(safePoint uint64) {
+		ticker := time.NewTicker(pauseGCIntervalForDupeRes)
+		defer ticker.Stop()
```

Comment:
I still think we need to remove the service gc safepoint after this function end. maybe by add a `defer pdCli.UpdateServiceGCSafePoint(ctx, serviceID, 0, safePoint)` . Or at lease, we should decrease the ttl to maybe 10min or shorter, 1 hour is too long

Reply:
I suggest removing it after finishing.

Reply:
Oh, I forgot it. 😂

Reply:
We need an enough ttl to tolerate lightning crash, otherwise duplicate records may be lost.

Reply:
@glorv  `safePoint` is removed after finishing now. PTAL again